calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -351,46 +358,108 @@ func (rh resourceHelper) GetObjectType() reflect.Type {
 // the typed interface on the client.
-func (rh resourceHelper) Create(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.create(client, resource)
+func (rh resourceHelper) Create(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	return rh.create(client, resource, ns)
 }
 
 // Update is an un-typed method to update an existing resource.  This calls directly
 // through to the resource helper specific Update method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Update(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.update(client, resource)
+func (rh resourceHelper) Update(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	for i := 0; i < 5; i++ {
+		ro, err := rh.GetOrList(client, resource, ns, false)
+		if err != nil {
+			return ro, err
+		}
+		if resource.GetObjectMeta().GetName() != "" {
+			resource.GetObjectMeta().SetResourceVersion(ro.(ResourceObject).GetObjectMeta().GetResourceVersion())
+		} else {
+			resource.GetObjectMeta().SetResourceVersion(ro.(ResourceListObject).GetListMeta().GetResourceVersion())
+		}
+
+		ru, err := rh.update(client, resource, ns)
+		if err != nil {
+			switch err.(type) {
+			// Wait for a second and try again if there was a conflict during the resource update.
+			case cerrors.ErrorResourceUpdateConflict:
+				log.Warnf("Error updating the resource %s: %s. Retrying.", resource.GetObjectMeta().GetName(), err)
+				time.Sleep(1 * time.Second)
+				continue
+
+			// For any other errors, return the error
+			default:
+				return ru, err
+			}
+		}
+
+		return ru, err
+	}
+	return nil, errors.New("failed to update the resource after 5 retries")
 }
 
 // Delete is an un-typed method to delete an existing resource.  This calls directly
 // through to the resource helper specific Delete method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Delete(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.delete(client, resource)
+func (rh resourceHelper) Delete(client client.Interface, resource ResourceObject, ns string) (runtime.Object, error) {
+	return rh.delete(client, resource, ns)
 }
 
 // Get is an un-typed method to get an existing resource.  This calls directly
 // through to the resource helper specific Get method which will map the untyped call to
 // the typed interface on the client.
-func (rh resourceHelper) Get(client client.Interface, resource ResourceObject) (ResourceObject, error) {
-	return rh.get(client, resource)
-}
+func (rh resourceHelper) GetOrList(client client.Interface, resource ResourceObject, ns string, allNs bool) (runtime.Object, error) {
+	if resource.GetObjectMeta().GetName() != "" {
+		if allNs {
+			return nil, fmt.Errorf("cannot use --all-namespace flag for getting a single resource")
```

Comment:
what does kubectl do if you use --all-namespaces to get a non-namespaces resource?  I assume it just ignores it?

Reply:
seems like they return an error like this:
```
$ kubectl get clusterrolebinding --all-namespaces
NAMESPACE   NAME      AGE
error: clusterRoleBinding is not namespaced
```